test/e2e/puppeteer.js

```diff
@@ -1,373 +1,628 @@
+
+					const renderStart = performance._now();
+
+					const waitingLoop = setInterval( function () {
+
+						const renderTimeoutExceeded = ( renderTimeout > 0 ) && ( performance._now() - renderStart > 1000 * renderTimeout );
+
+						if ( renderTimeoutExceeded ) {
+
+							clearInterval( waitingLoop );
+							reject( 'Render timeout exceeded' );
+
+						} else if ( window._renderFinished ) {
+
+							clearInterval( waitingLoop );
+							resolve();
+
+						}
+
+					}, 10 );
+
+				} );
+
+			}, renderTimeout, page.pageSize / 1024 / 1024 * parseTime * 1000 );
+
+		} catch ( e ) {
+
+			if ( e.message.includes( 'Render timeout exceeded' ) ) { // hack, TODO: remove this
+
+				console.yellow( `Render timeout exceeded in file ${ file }` );
+
+			} else {
+
+				throw new Error( `Error happened while rendering file ${ file }: ${ e }` );
 
-	if ( failedScreenshots.length ) {
+			}
+
+		}
+
+		const screenshot = await jimp.read( temporaryWebGPUHack ? await fromSurface( page ) : await page.screenshot() );
+		screenshot.scale( 1 / viewScale );
+
+		if ( page.error !== undefined ) throw new Error( page.error );
+
+		if ( isMakeScreenshot ) {
 
-		if ( failedScreenshots.length > 1 ) {
+			/* Make screenshot */
 
-			console.red( 'List of failed screenshots: ' + failedScreenshots.join( ' ' ) );
+			await screenshot.writeAsync( `examples/screenshots/${ file }.png` );
+			console.green( `Screenshot generated for file ${ file }` );
 
 		} else {
 
-			console.red( `If you sure that all is right, try to run \`npm run make-screenshot ${ failedScreenshots[ 0 ] }\`` );
+			/* Diff screenshots */
+
+			let expected;
+
+			try {
+
+				expected = await jimp.read( `examples/screenshots/${ file }.png` );
+
+			} catch {
+
+				await screenshot.writeAsync( `examples/output-screenshots/${ file }-actual.png` );
+				throw new Error( `Screenshot does not exist: ${ file }` );
+
+			}
+
+			const actual = screenshot.bitmap;
+			const diff = screenshot.clone();
+
+			let numFailedPixels;
+
+			try {
+
+				numFailedPixels = pixelmatch( expected.bitmap.data, actual.data, diff.bitmap.data, actual.width, actual.height, {
+					threshold: pixelThreshold,
+					alpha: 0.2,
+					diffMask: process.env.FORCE_COLOR === '0',
+					diffColor: process.env.FORCE_COLOR === '0' ? [ 255, 255, 255 ] : [ 255, 0, 0 ]
+				} );
+
+			} catch {
+
+				await screenshot.writeAsync( `examples/output-screenshots/${ file }-actual.png` );
+				await expected.writeAsync( `examples/output-screenshots/${ file }-expected.png` );
+				throw new Error( `Image sizes does not match in file: ${ file }` );
+
+			}
+
+			numFailedPixels /= actual.width * actual.height;
+
+			/* Print results */
+
+			if ( numFailedPixels >= 0.001 / 2 ) {
+
+				await screenshot.writeAsync( `examples/output-screenshots/${ file }-actual.png` );
+				await expected.writeAsync( `examples/output-screenshots/${ file }-expected.png` );
+				await diff.writeAsync( `examples/output-screenshots/${ file }-diff.png` );
```

Comment:
The screenshots are saved only for examples with non-zero diffs. And because GitHub Actions cleares the entire folder, I don't think we will run out of space...